### PR TITLE
Episode 17

### PR DESCRIPTION
**Pull request by bitsondatadev**

### CODE REVIEW

File: _episodes/17.md [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+If you want to show us some 💕, please [give us a ⭐ on GitHub](/star).
+
+<p align="center">
+<img align="center" width="50%" height="100%" src="/assets/episode/17/trino-resurface.png"/><br/>
+Commander Bun Bun is diving deep to find anomalies!
+</p>
+
+## Resurface links
+
+ * [Resurface Site](https://resurface.io/)
```
Comment (mosabua) [inline _episodes/17.md:40]: site

File: _episodes/17.md [hunk 2]
```diff
@@ -0,0 +1,416 @@
...
+ * [Resurface Github](https://github.com/resurfaceio)
```
Comment (mosabua) [inline _episodes/17.md:41]: GitHub

File: _episodes/17.md [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+ * [Resurface Slack](https://resurface.io/slack)
+
+## Guests
+ * Rob Dickinson, Co-founder and CEO of [Resurface Labs](https://resurface.io/)
```
Comment (mosabua) [inline _episodes/17.md:45]: is it Resurface Labs or just Resurface .. in the testimonial its just Resurface

File: _episodes/17.md [hunk 4]
```diff
@@ -0,0 +1,416 @@
...
+Resurface is an API system of record, which is a fancy way of saying that 
+Resurface is a purpose-built database for API requests and responses. Like a 
+weblog or access log, but on steroids because Resurface runs on Trino.
+
+Why do you need a system of record for your APIs? Because otherwise you’re 
+guessing about how your APIs are used and attacked, and guessing doesn’t feel 
+good. Resurface helps your DevOps and security teams instantly find API 
+failures, slowdowns, and attacks -- easily, responsibly, and at scale.
+
+### How Resurface Differs from Logs & Metrics
```
Comment (mosabua) [inline _episodes/17.md:62]: differs from logs and metrics

File: _episodes/17.md [hunk 5]
```diff
@@ -0,0 +1,416 @@
...
+
+<p align="center">
+<img align="center" width="75%" height="100%" src="/assets/episode/17/resurface-tcb1.png"/><br/>
+</p>
+
+The other obvious way to compare Resurface against other tools is to look at the
+data model. System monitoring gives you time-series metrics, or timestamped log
+messages with a severity and detail string. Resurface gives you all the request
+and response data fields, including headers and payloads, in a schema where all
+of those fields are discrete and searchable. Plus we add a bunch of helpful
```
Comment (mosabua) [inline _episodes/17.md:80]: And its adds a bunch

File: _episodes/17.md [hunk 6]
```diff
@@ -0,0 +1,416 @@
...
+virtual and computed columns.
+
+<p align="center">
+<img align="center" width="75%" height="100%" src="/assets/episode/17/resurface-tcb2.png"/><br/>
+</p>
+
+### The Indexing Problem
```
Comment (mosabua) [inline _episodes/17.md:87]: The indexing problem

File: _episodes/17.md [hunk 7]
```diff
@@ -0,0 +1,416 @@
...
+
+So we have this very descriptive data model, but there’s a problem here -- how
+to partition and index this data for efficient searching. Partitioning based on
+time is the obvious starting point, but within a time range, what then? Index
+everything?
+
+Most databases work best when a subset of the columns are constrained at once --
```
Comment (mosabua) [inline _episodes/17.md:94]: Change the whole post to not use "we" since this is TCB and not the Resurface website.

Comment (mosabua) [reply _episodes/17.md:94]: Or alternatively add some sort of intro that clarifies this is what Resurface says

File: _episodes/17.md [hunk 8]
```diff
@@ -0,0 +1,416 @@
...
+The example below is pretty much as simple as this gets. I struggled to find 
+one of these queries that actually fits in a reasonable amount of space.
+
+<p align="center">
+<img align="center" width="75%" height="100%" src="/assets/episode/17/resurface-tcb5.png"/><br/>
+</p>
+
+So how to build a database that does these kinds of queries in reasonable time?
+
+### The Resurface Connector
```
Comment (mosabua) [inline _episodes/17.md:125]: connector

File: _episodes/17.md [hunk 9]
```diff
@@ -0,0 +1,416 @@
...
+The connector was easy to write because we’re just mapping between fields in the
+binary files and the columns exposed to Trino. We built the first version of our
+connector in a weekend!
+
+<p align="center">
+<img align="center" width="75%" height="100%" src="/assets/episode/17/resurface-tcb3.png"/><br/>
+</p>
+
+### Why not just use Avro?
+ - Simple requirements -- basic versioning, no secondary objects, limited data
```
Comment (mosabua) [inline _episodes/17.md:152]: No indent and use * for lists

File: _episodes/17.md [hunk 10]
```diff
@@ -0,0 +1,416 @@
...
+# |Christeen|C623 |
+# |Christian|C623 |
+# |Christine|C623 |
+# |Chris    |C620 |
+# |Kristine |K623 |
+
+```
+
+
+## Question of the week: How to export query results into a file (e.g. CTAS, but into a single file)?
```
Comment (mosabua) [inline _episodes/17.md:322]: Empty line after title

File: _episodes/17.md [hunk 11]
```diff
@@ -0,0 +1,416 @@
...
+> CSV_HEADER, CSV_UNQUOTED, CSV_HEADER_UNQUOTED, JSON, NULL] (default: CSV)
+
+Here is an example of the command you would run using the cli executable `trino`.
+```
+trino --execute "select * from tpch.sf1.customer limit 5" \
+--server http://localhost:8080 \
+--output-format CSV_HEADER > customer.csv
+```
+
+If you're running Trino in docker, here is an example command to run this in a
```
Comment (mosabua) [inline _episodes/17.md:338]: Docker

File: _episodes/17.md [hunk 12]
```diff
@@ -0,0 +1,416 @@
...
+temporary trino container.
```
Comment (mosabua) [inline _episodes/17.md:339]: Trino

File: _episodes/17.md [hunk 13]
```diff
@@ -0,0 +1,416 @@
...
+This query along with the following `query.sql` file produces an equivalent query:
+```
+select * 
+from tpch.sf1.customer 
+limit 5;
+```
+
+Finally, one last trick is to stage the data using the memory connector to stage
+the data and finally export it. The [Trino Definitive Guide](https://www.starburst.io/info/oreilly-trino-guide/) 
+has example for adding iris data set into memory connector storage with cli.
```
Comment (mosabua) [inline _episodes/17.md:370]: the Iris data set .. with the CLI.

**Review by mosabua — CHANGES_REQUESTED** [review]
need to figure out voice and rejigger accordingly

File: _episodes/17.md (hunk 7 above)
Comment (bitsondatadev) [reply _episodes/17.md:94]: I went with the first one

**Review by mosabua — APPROVED** [review]
Looks good now.

File: _episodes/17.md [hunk 14]
```diff
@@ -0,0 +1,417 @@
...
+Trino Meetup groups
+ - Virtual
+   - [Virtual Trino Americas](https://www.meetup.com/trino-americas/)
+   - [Virtual Trino EMEA](https://www.meetup.com/trino-emea/)
+   - [Virtual Trino APAC](https://www.meetup.com/trino-apac/)
+ - East Coast (US)
+   - [Trino Boston](https://www.meetup.com/trino-boston/)
+   - [Trino NYC](https://www.meetup.com/trino-nyc/)
+ - West Coast (US)
+   - [Trino San Fransisco](https://www.meetup.com/trino-san-francisco/)
```
Comment (martint) [inline _episodes/17.md:400]: Francisco

**Pull request merged**